packages/datepicker-react/src/internal/reducer.ts

```diff
@@ -0,0 +1,54 @@
+import { formatDate, isSameDay, parseDateString } from "../dateFunctions";
+
+interface SetDate {
+    type: "SET_DATE";
+    payload: Date | undefined;
+}
+
+interface SetDateString {
+    type: "SET_DATE_STRING";
+    payload: string | undefined;
+}
+
+interface ToggleCalendar {
+    type: "TOGGLE_CALENDAR";
+    payload: boolean;
+}
+
+export type Action = SetDate | SetDateString | ToggleCalendar;
+
+export interface State {
+    date?: Date;
+    dateString?: string;
+    calendarHidden: boolean;
+}
+
+const dateHasChanged = (date: Date | undefined, newDate: Date) => {
+    return !date || !isSameDay(date, newDate);
+};
+
+export const reducer = (state: State, action: Action): State => {
+    switch (action.type) {
+        case "SET_DATE":
+            return { ...state, date: action.payload, dateString: action.payload ? formatDate(action.payload) : "" };
+
+        case "SET_DATE_STRING":
+            let date = state.date;
+            const newDate = parseDateString(action.payload ?? "");
+
+            // Only set the date if it is a valid date
+            if (newDate && dateHasChanged(date, newDate)) {
+                date = newDate;
+            } else if (action.payload === "") {
+                date = undefined;
+            }
+
+            return { ...state, dateString: action.payload, date };
+
+        case "TOGGLE_CALENDAR":
```

Comment:
hvis actionen er TOGGLE, så burde den vel toggle, ikke sette verdien fra payload?

`calendarHidden: !state.calendarHidden`

Reply:
Ja, det gir mening. Jeg tenker det fortsatt er best å la payload sette statusen, så jeg renamer den heller til `SET_CALENDAR_OPEN`. Dette for å gi muligheten til at det ikke alltid er en toggle, og kan gi uønsket oppførsel ved at toggle state kommer ut av sync med resten av logikken.

Reply:
Enig, evt begge deler, en TOGGLE og en SET